src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,283 @@
+    // ABIs (incl. intrinsics).
+    let node = tcx.hir().get(hir_id);
+    debug!("used_generic_parameters: node={:?}", node);
+    match node {
+        hir::Node::Expr(&hir::Expr { kind: hir::ExprKind::Closure(..), .. }) => (),
+        hir::Node::Item(&hir::Item { kind: hir::ItemKind::Fn(..), .. })
+        | hir::Node::ImplItem(&hir::ImplItem { kind: hir::ImplItemKind::Method(..), .. })
+        | hir::Node::TraitItem(&hir::TraitItem {
+            kind: hir::TraitItemKind::Method(_, hir::TraitMethod::Provided(..)),
+            ..
+        }) => match tcx.fn_sig(def_id).abi() {
+            Abi::Rust => (),
+            _ => return BitSet::new_filled(generics.count()),
+        },
+        _ => return BitSet::new_filled(generics.count()),
+    }
+
+    let mut used_parameters = BitSet::new_empty(generics.count());
+    debug!(
+        "used_generic_parameters: used_parameters.domain_size()={:?}",
+        used_parameters.domain_size()
+    );
+    if generics.count() == 0 {
+        // Exit early when there are no parameters to be unused.
+        return used_parameters;
+    }
+
+    let leaf_is_closure =
+        if let hir::Node::Expr(&hir::Expr { kind: hir::ExprKind::Closure(..), .. }) = node {
+            true
+        } else {
+            false
+        };
+
+    mark_used_by_default_parameters(tcx, def_id, generics, leaf_is_closure, &mut used_parameters);
+
+    // Visit MIR and accumululate used generic parameters.
+    let body = tcx.optimized_mir(def_id).unwrap_read_only();
+    let mut vis =
+        UsedGenericParametersVisitor { tcx, def_id, body, used_parameters: &mut used_parameters };
+    vis.visit_body(body);
+    debug!("used_generic_parameters: used_parameters={:?}", used_parameters);
+
+    // Emit errors for debugging and testing if enabled.
+    let is_full = used_parameters.count() == used_parameters.domain_size();
+    if tcx.sess.opts.debugging_opts.polymorphize_errors && !is_full {
+        emit_used_generic_parameters_error(tcx, def_id, generics, &used_parameters);
+    }
+
+    used_parameters
+}
+
+/// Some parameters are considered used-by-default, such as non-generic parameters and the dummy
+/// generic parameters from closures, this function marks them as used. `leaf_is_closure` should
+/// be `true` if the item that `used_generic_parameters` was invoked on is a closure.
+fn mark_used_by_default_parameters<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    generics: &'tcx ty::Generics,
+    leaf_is_closure: bool,
+    used_parameters: &mut BitSet<u32>,
+) {
+    for param in &generics.params {
+        debug!("mark_used_by_default_parameters: param={:?}", param);
+        // Consider any of the dummy generic parameters that closures have as used.
+        if leaf_is_closure && param.def_id == def_id {
+            used_parameters.insert(param.index);
+        }
+
+        // Consider every non-generic parameter used by default.
```

Comment:
These are all generic parameters, so the comment is confusing.
Also, it's probably better to special-case lifetimes as the ones to mark as used (although it's not necessary I don't think, since you do check the type of parameter when you use this bitset later).
As for `const` parameters, you need to treat those like type parameters, as uniformly as possible.

Reply:
I've fixed this by special-casing lifetimes only, and supporting const parameters in 4dc7f3f.